apps/designer/app/designer/shared/tree/index.tsx

```diff
@@ -55,8 +57,22 @@ const instanceRelatedProps = {
 } as const;
 
 export const InstanceTree = (
-  props: Omit<TreeProps<Instance>, keyof typeof instanceRelatedProps>
-) => <Tree {...props} {...instanceRelatedProps} />;
+  props: Omit<
+    TreeProps<Instance>,
+    keyof typeof instanceRelatedProps | "findItemById"
+  >
+) => {
+  const { instancesById } = useStore(instancesIndexStore);
+  const findItemById = useCallback(
+    (_rootInstance: Instance, instanceId: Instance["id"]) => {
```

Comment:
Looking at `_rootInstance` here makes me realise that `TreeProps<Data>` is a bit wrong. It should be like this:

```js
export type TreeProps<Root, Data extends { id: string }> = {
  root: Root;

  canLeaveParent: (item: Data) => boolean;
  canAcceptChild: (item: Data) => boolean;
  findItemById: (root: Root, id: string) => Data | undefined;
  getItemPath: (root: Root, id: string) => Data[];
  getItemPathWithPositions: (
    root: Root,
    id: string
  ) => Array<{ item: Data; position: number }>;
  getItemChildren: (item: Data) => Data[];
  renderItem: (props: TreeItemRenderProps<Data>) => React.ReactNode;

  selectedItemId?: string;
  onSelect?: (itemId: string) => void;
  onHover?: (item: Data | undefined) => void;
  animate?: boolean;
  onDragEnd: (event: {
    itemId: string;
    dropTarget: { itemId: string; position: number | "end" };
  }) => void;
};
```

Then you could have whatever as root, in this case I'd make `{rootInstance, instancesById}` the root object. And then all these functions like `findItemById` can be defined outside of render again.

Up to you whether to refactor this.

Reply:
Will refactor this way when figure out all case